internal/app/init/internal/reg/reg.go

```diff
@@ -62,3 +63,22 @@ func (r *Registrator) Shutdown(ctx context.Context, in *empty.Empty) (reply *pro
 
 	return
 }
+
+// Upgrade initiates a Talos upgrade
+func (r *Registrator) Upgrade(ctx context.Context, in *proto.UpgradeRequest) (data *proto.UpgradeReply, err error) {
+
+	if err = upgrade.NewUpgrade(in.Url); err != nil {
+		return data, err
+	}
+
+	// Trigger reboot
+	defer func() {
+		if _, err = r.Reboot(ctx, &empty.Empty{}); err != nil {
+			return
+		}
+	}()
+
+	// profit
+	data = &proto.UpgradeReply{Ack: "Upgrade initiated"}
```

Comment:
nit: the upgrade is complete at this point, i suggest `Ack: "upgrade complete, rebooting node"`